### DISCUSSION TITLE
Astro.scope

### DISCUSSION THREAD

**Discussion by lilnasy** [withastro/roadmap#706] [Proposal]
# Summary
Expose to userland the compiler-generated hash currently used for scoping css rules.

# Example
```astro
---
const scope : string = Astro.scope
---
```

# Goals
- Allow astro components to share scope.
- Allow client-side javascript to select elements authored in the same component.

# Harmony
_How might this proposal interact with other in-progress proposals?_
- https://github.com/withastro/roadmap/discussions/705 will allow Astro.scope to be passed to client-side scripts.
```html
<h1>Welcome to Astro</h1>
<h3>The web framework that scales with you</h3>

<script define:args={Astro.scope}>
    let selectedAlready = false
    
    export default function select (scope : string) {
        // this function needs to run only once per page load
        if (selectedAlready) { return }
        else { selectedAlready = true }

        // all instances of elements authored in the current component
        const elements = document.querySelector('[data-astro-cid-' + scope +']')
    }
</script>
```

**Comment by matthewp** [comment]
One question I have is if the scope is really what people want. The use-case in the example (and the one I hear most about) is wanting to be able to query for the component's elements in a client script. In that case the scope isn't ideal because it's going to match every element in the component, but also every *instance* of the component. Would something like this work better?

```astro
---
const id = Astro.uniqueId();
---

<h1 id={id}>Welcome to Astro</h1>
<h3>The web framework that scales with you</h3>

<script define:args={id}>
    export default function select (scope : string) {
        // just the h1
        const h1 = document.getElementById(id)
    }
</script>
```

Some pros to this approach:

- It's just an id, you can use it however you like, attach to some elements or all of them, as any attribute you like.
- Will be unique to the component instance.
- IIRC, the scope is not always on every element in an astro component, we only add that attribute to elements that have styles. So you'd still need to attach scope to an attribute manually if you don't have styles. This seems like it could be confusing.

**Reply by lilnasy** [reply]
This is definitely better.

Users might have to add a class name to every element on the page to select everything rendered by the component, but I'm not sure if there's a use case there.

**Reply by openscript** [reply]
I think something like this should be added to Astro. Should I send a PR for this?

**Reply by florian-lefebvre** [reply]
This kind of proposal needs to go through the whole rfc process so do not submit a pr yet! Looking at the other comments, I don't think there's a clear consensus so i don't think this will be tackled shortly. In the meantime, you can use the scope package from Arsh

**Reply by openscript** [reply]
After discussing this in Discord, I think it will lead to rendering duplicated Javascript. I guess in most of the cases there are alternatives, like making the script working globally. For example `closest` becomes very useful when you need to select a specific parent or sibling when an event happens.

To proceed with this, I would suggest to collect examples where is no other choice than having such IDs. After all, it's easy to implement your own ID if you need it.

**Reply by delucis** [reply]
It seems very often custom elements are a good answer here where you want to run JS for the “current” component instance. That’s what they’re designed for (amongst other things).

https://docs.astro.build/en/guides/client-side-scripts/#web-components-with-custom-elements

**Comment by wassfila** [comment]
can this somehow be used to scope e.g. markdown from Post.Content, or html set with set:html ? cause that would be a big releaf from having to perform manual scoping with long class names.

**Reply by lilnasy** [reply]
Yes, although you would need to add the scope to classes (or attributes) manually.

**Comment by natemoo-re** [comment]
Hey—@bholmesdev made a neat Vite plugin that exposes something similar. We should steal his great idea!

https://www.npmjs.com/package/vite-plugin-simple-scope

**Reply by lilnasy** [reply]
Done!
[GitHub](https://github.com/lilnasy/gratelets/tree/main/packages/scope) | [NPM](https://www.npmjs.com/package/astro-scope)
